File: tests/runners/csharp/tests/Test_04_inline_enums.cs
// Test Case 04: Inline Enums
// Tests enum definitions inside tables

using System;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string message)
    {
        if (!condition)
        {
            Console.WriteLine($"    FAILED: {message}");
            failed++;
        }
    }

    static void TestOrderInlineEnums()
    {
        Console.WriteLine("  Testing Order inline enums...");

        var order = new test.enums.Order
        {
            id = 1,
            customer_name = "John Doe",
            status = test.enums.Order.Status.Pending,
            priority = test.enums.Order.Priority.High
        };

        Assert(order.id == 1, "id");
        Assert(order.customer_name == "John Doe", "customer_name");
        Assert(order.status == test.enums.Order.Status.Pending, "status");
        Assert(order.priority == test.enums.Order.Priority.High, "priority");

        // Test all status values
        order.status = test.enums.Order.Status.Paid;
        Assert((int)order.status == 1, "Paid value");

        order.status = test.enums.Order.Status.Shipped;
        Assert((int)order.status == 2, "Shipped value");

        order.status = test.enums.Order.Status.Delivered;
        Assert((int)order.status == 3, "Delivered value");

        order.status = test.enums.Order.Status.Cancelled;
        Assert((int)order.status == 4, "Cancelled value");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestTaskInlineEnum()
    {
        Console.WriteLine("  Testing Task inline enum...");

        var task = new test.enums.Task
        {
            id = 1,
            title = "Complete tests",
            state = test.enums.Task.State.Todo
        };

        Assert(task.state == test.enums.Task.State.Todo, "Todo");

        task.state = test.enums.Task.State.InProgress;
        Assert((int)task.state == 1, "InProgress value");

        task.state = test.enums.Task.State.Done;
        Assert((int)task.state == 2, "Done value");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestGlobalEnum()
    {
        Console.WriteLine("  Testing global enum (GlobalStatus)...");

        var status = test.enums.GlobalStatus.Unknown;
        Assert((int)status == 0, "Unknown value");

        status = test.enums.GlobalStatus.Active;
        Assert((int)status == 1, "Active value");

        status = test.enums.GlobalStatus.Disabled;
        Assert((int)status == 2, "Disabled value");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestBinaryInlineEnums()
    {
        Console.WriteLine("  Testing binary serialization with inline enums...");

        var original = new test.enums.Order
        {
            id = 12345,
            customer_name = "Test Customer",
            status = test.enums.Order.Status.Shipped,
            priority = test.enums.Order.Priority.Urgent
        };

        // Serialize
        using var ms = new System.IO.MemoryStream();
        using var writer = new System.IO.BinaryWriter(ms);
        test.enums.BinaryWriters.WriteOrder(writer, original);

        // Deserialize
        ms.Position = 0;
        using var reader = new System.IO.BinaryReader(ms);
        var loaded = test.enums.BinaryReaders.ReadOrder(reader);

        Assert(loaded.id == original.id, "id mismatch");
        Assert(loaded.customer_name == original.customer_name, "customer_name mismatch");
        Assert(loaded.status == original.status, "status mismatch");
        Assert(loaded.priority == original.priority, "priority mismatch");

        Console.WriteLine($"    PASS (serialized {ms.Length} bytes)");
        passed++;
    }

    static void Main()
    {
        Console.WriteLine("=== Test Case 04: Inline Enums ===");

        TestOrderInlineEnums();
        TestTaskInlineEnum();
        TestGlobalEnum();
        TestBinaryInlineEnums();

        if (failed > 0)
        {
            Console.WriteLine($"=== {failed} tests failed! ===");
            Environment.Exit(1);
        }
        else
        {
            Console.WriteLine("=== All tests passed! ===");
        }
    }
}

File: tests/runners/cpp/tests/test_04_inline_enums.cpp
// Test Case 04: Inline Enums
// Tests enum definitions inside tables

#include <iostream>
#include <cassert>

#include "schema.hpp"
#include "schema_loaders.hpp"

using namespace test::enums;

void test_order_inline_enums() {
    std::cout << "  Testing Order inline enums..." << std::endl;

    Order order;
    order.id = 1;
    order.customer_name = "John Doe";
    order.status = Order::Status::Pending;
    order.priority = Order::Priority::High;

    assert(order.id == 1);
    assert(order.customer_name == "John Doe");
    assert(order.status == Order::Status::Pending);
    assert(order.priority == Order::Priority::High);

    // Test all status values
    order.status = Order::Status::Paid;
    assert(static_cast<int32_t>(order.status) == 1);

    order.status = Order::Status::Shipped;
    assert(static_cast<int32_t>(order.status) == 2);

    order.status = Order::Status::Delivered;
    assert(static_cast<int32_t>(order.status) == 3);

    order.status = Order::Status::Cancelled;
    assert(static_cast<int32_t>(order.status) == 4);

    std::cout << "    PASS" << std::endl;
}

void test_task_inline_enum() {
    std::cout << "  Testing Task inline enum..." << std::endl;

    Task task;
    task.id = 1;
    task.title = "Complete tests";
    task.state = Task::State::Todo;

    assert(task.state == Task::State::Todo);

    task.state = Task::State::InProgress;
    assert(static_cast<int32_t>(task.state) == 1);

    task.state = Task::State::Done;
    assert(static_cast<int32_t>(task.state) == 2);

    std::cout << "    PASS" << std::endl;
}

void test_global_enum() {
    std::cout << "  Testing global enum (GlobalStatus)..." << std::endl;

    GlobalStatus status = GlobalStatus::Unknown;
    assert(static_cast<int32_t>(status) == 0);

    status = GlobalStatus::Active;
    assert(static_cast<int32_t>(status) == 1);

    status = GlobalStatus::Disabled;
    assert(static_cast<int32_t>(status) == 2);

    std::cout << "    PASS" << std::endl;
}

void test_binary_inline_enums() {
    std::cout << "  Testing binary serialization with inline enums..." << std::endl;

    Order original;
    original.id = 12345;
    original.customer_name = "Test Customer";
    original.status = Order::Status::Shipped;
    original.priority = Order::Priority::Urgent;

    // Serialize
    std::vector<uint8_t> buffer;
    {
        polygen::BinaryWriter writer(buffer);
        polygen_loaders::write_Order(writer, original);
    }

    // Deserialize
    Order loaded;
    {
        polygen::BinaryReader reader(buffer);
        loaded = polygen_loaders::read_Order(reader);
    }

    assert(loaded.id == original.id);
    assert(loaded.customer_name == original.customer_name);
    assert(loaded.status == original.status);
    assert(loaded.priority == original.priority);

    std::cout << "    PASS (serialized " << buffer.size() << " bytes)" << std::endl;
}

int main() {
    std::cout << "=== Test Case 04: Inline Enums ===" << std::endl;

    test_order_inline_enums();
    test_task_inline_enum();
    test_global_enum();
    test_binary_inline_enums();

    std::cout << "=== All tests passed! ===" << std::endl;
    return 0;
}
